fix(client): 🐛 Correct excessive writes to firestore

gchamb opened the pull request:
- fix excessive writes to firestore by switching on change event to on blur.
- removed user logic from buttons
- renamed function and return type

#26

gchamb committed:
fix(client): 🐛 Correct excessive writes to firestore

- fix excessive writes to firestore by switching onchange event to on blur.
- removed user logic from buttons
- renamed function and return type

nstringham commented on client/src/lib/firebase/splash.ts:
```diff
@@ -6,9 +6,9 @@ import { doc, updateDoc, getDoc, setDoc } from "firebase/firestore";
+export const getUser = async (id: string) => {
   const user = await getDoc(doc(db, `/users/${id}`));
-  return user.data() as UserData;
+  return user.data() as UserData | undefined;
```
Isn't `UserData | undefined` already the return type of data()?

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -1,45 +1,74 @@
+    }
+    // this is an anon user
+    // create anon user and user doc with display name
+    if (user === null && userData === undefined) {
```
What if the user has logged in after the onmount was called?

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -85,10 +114,10 @@
     </div>
     <div class="cat-main-buttons">
-      <input type="text" placeholder="Enter in your display name" on:change={saveName} bind:value={name} />
+      <input type="text" placeholder="Enter in your display name" on:blur={saveOrCreate} bind:value={name} />
```
Might also consider `oninput` not sure which is better

nstringham reviewed (changes requested):
Could also consider creating the user document when the use clicks the create/join lobby button

gchamb committed:
Merge branch 'development' into 26-splash-screen

gchamb committed:
fix: merge stashed changes

gchamb committed:
refactor: removed console log in tests

gchamb committed:
feat(client): ✨ Finish Firestore Rule for Users

gchamb committed:
Merge branch 'development' into 26-splash-screen

nstringham commented on firestore.rules:
```diff
@@ -8,19 +8,17 @@ service cloud.firestore {
+        allow read;
+        allow create, update: if request.auth.uid == userId && validateAttrs();
+        allow delete: if request.auth.uid == userId;
 
```
Are we sure we want users to be able to delete this?

nstringham commented on firestore.rules:
```diff
@@ -8,19 +8,17 @@ service cloud.firestore {
-        // request.resource.data.displayName.match('/[0-9A-Za-z_-]/')
-        request.resource.data.avatar is number && request.resource.data.avatar >= 0 && request.resource.data.avatar <=12
+        request.resource.data.displayName.matches('[A-Za-z0-9-_]+') &&
+        int(request.resource.data.avatar) is number && request.resource.data.avatar >= 0 && request.resource.data.avatar <=12
```
Wouldn't this allow strings because we are casting to int?

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -1,49 +1,76 @@
+    // Create User
+    saveOrCreate();
+    // Creat Lobby
+    const code = await createLobby();
```
This seams like a race condition to me between user and game getting created

gchamb committed:
fix(client): 🐛 Fix firestore rule and race condition

gchamb committed:
Merge branch 'development' into 26-splash-screen

gchamb committed:
Merge branch '26-splash-screen' of https://github.com/cat-acronym-team/cat-me-if-you-can into 26-splash-screen

gchamb committed:
refactor: Replaced with async await

- replaced then and catch with await

gchamb committed:
feat(client): ✨ Include RX Fire to subscribe auth

- added rx fire
- subscribes the user to rxfire observable
- changed tests to match firebase rules

gchamb committed:
fix: remove rxfire from base to client package.json

gchamb committed:
feat(client): ✨ Introduce auth syncing

- removed rxfire
- created auth store
- removed onmount for a reactive function to have display name update once auth isn't null
- added a new alias

gchamb committed:
Merge branch 'development' into 26-splash-screen

nstringham commented on client/src/lib/firebase/app.ts:
```diff
@@ -37,6 +38,9 @@ const app = initializeApp(firebaseConfig);
 export const db = getFirestore(app);
 export const auth = getAuth(app);
 
+auth.onAuthStateChanged((user) => {
+  authStore.set(user);
+});
```
will this code be run every time this component is loaded?
would it be better to put it in `$stores/auth`?

nstringham replied:
this was me being dumb because I thought this was +page.svelte for some reason

nstringham commented on client/svelte.config.js:
```diff
@@ -11,6 +11,7 @@ const config = {
     adapter: adapter(),
     alias: {
       $components: "./src/components",
+      $stores: "./src/store",
```
it seams like some people the internet use `$lib/stores` instead of `$stores` but either way seams fine to me

nstringham replied:
we decided we like `$stores`

nstringham commented on firestore.rules:
```diff
@@ -8,19 +8,16 @@ service cloud.firestore {
+      	request.resource.data.keys().hasOnly(['displayName', 'avatar']) &&
         request.resource.data.displayName is string && request.resource.data.displayName.trim().size() >= 3 && request.resource.data.displayName.trim().size() <= 15 &&
-        // TODO: finish display name regex
-        // request.resource.data.displayName.match('/[0-9A-Za-z_-]/')
-        request.resource.data.avatar is number && request.resource.data.avatar >= 0 && request.resource.data.avatar <=12
+        request.resource.data.displayName.matches('[A-Za-z0-9-_]+') &&
+        request.resource.data.avatar is number &&  int(request.resource.data.avatar) == request.resource.data.avatar &&  request.resource.data.avatar >= 0 && request.resource.data.avatar <=12
```
normally I would indent this

nstringham replied:
lets try to keep lines less than around 120 charicters long.

you can check by placing you're cusor at the end of the line and seeing that VS-Code says "Ln 14, Col 192" at the bottom. this means that you are on line 14 and coloumn 192 which means that the line is longer than 120 charicters

nstringham commented on package-lock.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "cat-me-if-you-can",
+  "name": "cat-project",
```
why?

gchamb replied:
I have it named as that on my machine so it looks like it named it that in the package-lock.json. Mistake.

nstringham replied:
hmm. maybe we should set a name in package.json to avoid this

nstringham reviewed (changes requested):
oops I made those comments not part of this review. sorry

gchamb committed:
feat(client): ✨ Unsubscribes from onAuthChanged

- fixed indent and lines
- created custom store instead of the base one
- added unsubscribe from onAuthChanged
- Splash unsubs store from onAuthChanged when destroyed

nstringham committed:
revert(client): ⏪️ don't Unsubscribes from onAuthChanged

because this was implemented because of a misunderstanding

Co-authored-by: Giantte Chambers <gchamb@users.noreply.github.com>

This partially reverts commit 8aff9f0a7202e59ae5209d7d86e3c9082f693a7c.

nstringham reviewed (approved):
code looks good.
tested and working as expected.
